Add parameter to control initial scatter point generation

DannyHindson opened the pull request:
**Description of work.**

This PR came out of an aborted attempt to change the logic used to sample the initial scatter point. I initially thought the sampling was invalid for shapes like cylinders or spheres where the depth varies across xy - but have now decided it's fine after all. There are a couple of remnants of the original (larger) code change though that are useful to retain:

- I've added a new parameter called MaxScatterPtAttempts to control the scatter point generation. This brings the algorithm into line with the MonteCarloAbsorption algorithm
- comments are now more clear
- I've changed one of the "reference" values in the unit tests to be better centred in the range calculated when using different random seeds. It was on the edge of the range before so the test only passed when using the default seed. Doing any code change (like the abandoned one) that changed the random number sequence caused the test to fail

**To test:**

Run this script:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from testhelpers import WorkspaceCreationHelper

test_ws = WorkspaceCreationHelper.create2DWorkspaceWithGeographicalDetectors(1, 900, 0.2, 1)
ws_name = "test"
AnalysisDataService.addOrReplace(ws_name, test_ws)

X=[0.5,1.5]
Y=[1.0]
Sofq=CreateWorkspace(DataX=X,DataY=Y,UnitX="MomentumTransfer")

sphere_xml = " \
<sphere id='some-sphere'> \
    <centre x='0.0'  y='0.0' z='0.0' /> \
    <radius val='0.001' /> \
</sphere> \
<algebra val='some-sphere' /> \
"

SetSample(InputWorkspace=test_ws,
          Geometry={'Shape': 'CSG', 'Value': sphere_xml},
          Material={'ChemicalFormula': 'Ni'})

DiscusMultipleScatteringCorrection(InputWorkspace=test_ws, SofqWorkspace=Sofq, OutputWorkspace="MSResults",
                                   NeutronPathsSingle=1, NeutronPathsMultiple=10000, NumberScatterings=2,
                                   MaxScatterPtAttempts=100)
```

If you set MaxScatterPtAttempts to 1 the algorithm will fail saying it can't generate a scatter point in the sample. This is because the algorithm uses the bounding box of the sample to generate scatter points.

Fixes #32766.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Add parameter to control scatter point generation

This aligns the algorithm with the MonteCarloAbsorption algorithm
I've also adjusted the comments near the scatter point generation in
GenerateInitialTrack to remove the doubt over whether the scatter point
sampling is correct

ConorMFinn commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -516,7 +525,7 @@ double DiscusMultipleScatteringCorrection::simulatePaths(const int nPaths, const
  * @param rng Random number generator
- * @param sigma_total
+ * @param sigma_total The total cross section (scattering + absorption)
  * @param scatteringXSection
```
Could this param also have a comment?

ConorMFinn commented on Framework/Algorithms/test/DiscusMultipleScatteringCorrectionTest.h:
```diff
@@ -175,10 +175,12 @@ class DiscusMultipleScatteringCorrectionTest : public CxxTest::TestSuite {
-      TS_ASSERT_DELTA(doubleScatterY[2], 0.001819, delta);
-      TS_ASSERT(doubleScatterY[1] < 0.001977 || doubleScatterY[1] < 0.001819);
-      TS_ASSERT(doubleScatterY[1] > 0.001977 || doubleScatterY[1] > 0.001819);
+      double expResult0 = 0.001977;
+      double expResult2 = 0.001810;
```
I think these can be made `constexpr`?

ConorMFinn commented on buildconfig/CMake/CppCheck_Suppressions.txt.in:
```diff
@@ -516,10 +516,10 @@ useStlAlgorithm:${CMAKE_SOURCE_DIR}/Framework/DataHandling/src/UpdateInstrumentF
-useStlAlgorithm:${CMAKE_SOURCE_DIR}/Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:257
 constParameter:${CMAKE_SOURCE_DIR}/Framework/Algorithms/src/CalculatePlaczekSelfScattering.cpp:29
 constParameter:${CMAKE_SOURCE_DIR}/Framework/Algorithms/src/CalculatePlaczek.cpp:45
+passedByValue:${CMAKE_SOURCE_DIR}/Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:491
+useStlAlgorithm:${CMAKE_SOURCE_DIR}/Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:266
```
Would it be possible/trivial to fix these instead of moving the suppressions?

cailafinn reviewed (changes requested):
Looks good and works well, just a couple of minor comments:

DannyHindson committed:
Couple of minor changes

Resolve cppcheck warning by using std::all_of and change constants
to be constexpr

ConorMFinn commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -249,9 +258,9 @@ void DiscusMultipleScatteringCorrection::exec() {
-  for (auto &ws : simulationWSs) {
-    enableParallelFor = enableParallelFor && Kernel::threadSafe(*ws);
-  }
+  enableParallelFor = std::all_of(simulationWSs.begin(), simulationWSs.end(),
+                                  [](MatrixWorkspace_sptr ws) { return Kernel::threadSafe(*ws); });
```
Could `cbegin()` and `cend()` be used here? Could `ws` also be a ref?

DannyHindson replied:
Yes I'll change to cbegin\cend. 
Re your other point - are you wondering if the lambda should have `MatrixWorkspace_sptr& ws` or `const MatrixWorkspace_sptr& ws`? Since ws is actually a pointer I'm not sure passing it by reference would add much?

ConorMFinn replied:
It's not a lot, but it would stop there being a copy made of the pointer, which is a small memory save if nothing else. I think. I'll admit I'm not the most clued up on when to use pointers/references/etc. 😅

DannyHindson replied:
Since this is a smart pointer I think you're right. Taking a copy of the smart pointer probably is more heavy than copying a memory location (which is what you get when passing a raw pointer or doing a pass by reference). I guess some reference counts perhaps gets incremented\decremented with the existing code as well which is further slowness. So will change!

ConorMFinn commented on buildconfig/CMake/CppCheck_Suppressions.txt.in:
```diff
@@ -518,7 +518,6 @@ nullPointer:${CMAKE_SOURCE_DIR}/Framework/Kernel/inc/MantidKernel/IPropertyManag
 funcArgOrderDifferent:${CMAKE_SOURCE_DIR}/Framework/Algorithms/src/CalculateEfficiency.cpp:308
 constParameter:${CMAKE_SOURCE_DIR}/Framework/Algorithms/src/CalculatePlaczekSelfScattering.cpp:29
 constParameter:${CMAKE_SOURCE_DIR}/Framework/Algorithms/src/CalculatePlaczek.cpp:45
-useStlAlgorithm:${CMAKE_SOURCE_DIR}/Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:253
```
One less of these is always nice!

cailafinn reviewed (changes requested):
Looks good, just a tiny note:

DannyHindson committed:
Use const iterators instead

DannyHindson committed:
Pass smart pointer by reference

cailafinn reviewed (approved):
Looks good to me! Nice one 👍🏻
